File: No2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class No2 {

    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
        }
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int plus = 0;
        ListNode result;
        ListNode p = new ListNode(-1);
        result = p;

        // first
        while(l1 != null && l2 != null) {
            ListNode tmp = new ListNode((l1.val + l2.val)%10 + plus);
            plus = l1.val + l2.val > 9 ? 1 : 0;
            p.next = tmp;
            p = tmp;
            l1 = l1.next;
            l2 = l2.next;
        }
        // second
        while(l1 != null) {
            ListNode tmp = new ListNode((plus + l1.val) % 10);
            plus = (plus + l1.val) / 10;
            p.next = tmp;
            p = tmp;
            l1 = l1.next;
        }
        while(l2 != null) {
            ListNode tmp = new ListNode((plus + l2.val) % 10);
            plus = (plus + l2.val) / 10;
            p.next = tmp;
            p = tmp;
            l1 = l2.next;
        }
        // third
        if (plus == 1) {
            ListNode tmp = new ListNode(1);
            p.next = tmp;
        }

        return result.next;
    }
    public ListNode addTwoNumbers_right(ListNode l1, ListNode l2) {
        ListNode prev = new ListNode(0);
        ListNode head = prev;
        int carry = 0;
        while(l1 != null || l2 != null || carry == 1) {
            int sum = (l1 == null ? 0 : l1.val) + (l2 == null ? 0 : l2.val) + carry;
            carry = sum / 10;
            sum = sum % 10;
            ListNode temp = new ListNode(sum);
            prev.next = temp;
            prev = temp;
            head.val++;

            l1 = (l1 == null ? null : l1.next);
            l2 = (l2 == null ? null : l2.next);
        }
        return head.next;
    }
}

File: python/No2.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        acc, mod = 0, 0
        p1, p2 = l1, l2
        rtn = ListNode()
        r = rtn
        while p1 or p2 or acc:
            sum = (p1.val if p1 else 0) + (p2.val if p2 else 0) + acc
            mod = sum-10 if sum>=10 else sum
            r = self._add_node(r, mod)
            acc = 1 if sum>=10 else 0
            p1 = p1.next if p1 else None
            p2 = p2.next if p2 else None
        
        return rtn.next
    
    def _add_node(self, head, val):
        new_node = ListNode(int(val), None)
        head.next = new_node
        return new_node
    def build_list_node(self, arr):
        head = ListNode()
        tmp = head
        for item in arr:
            tmp = self._add_node(tmp, int(item))
        return head.next
    def printf(self, head):
        tmp = []
        item = head
        while item:
            tmp.append(str(item.val))
            item = item.next
        print(' '.join(tmp))

c = Solution()
#A = c.build_list_node([9,9,9,9])
#B = c.build_list_node([9,9])
A = c.build_list_node([2,4,3])
B = c.build_list_node([5,6,4])
c.printf(A)
c.printf(B)
c.printf(c.addTwoNumbers(A, B))